docs/src/crypto-primitives/tsmt.md

```diff
@@ -0,0 +1,123 @@
+# Tiered Sparse Merkle Tree (TSMT)
+A sparse Merkle tree (SMT) is a cryptographic data structure that allows authenticated manipulations of a key-value store i.e. a dictionary. It is a Merkle tree because it is
+a binary tree in which each leaf node represents a key-value pair, and each non-leaf node represents the hash of the concatenation of its two child nodes. The index of the
+leaf node in the tree corresponds to the key and the data stored in this node corresponds to the value. This implies that the depth of such a tree is logarithmic in the size of the key space, which implies that, for most key spaces used in practice, such a tree is impractical if done naively. However, the tree is sparse i.e it is going to be made up of mostly empty leaves which in turn implies that most sub-trees are empty with roots that can be pre-computed. 
+A major issue with SMT is the fact that operations scale with the size of the key space meaning, for example, that if our key space has size $~ 2^{256}$ then performing any operation on the key-value store, for example an update, will require performing $256$ hashing operations.This is undesirable both inside as well as outside the STARK.
+
+
+A common solution to this issue is to keep only non-empty leaves in the tree, which in practice means that sub-trees with only one non-empty child are replaced with the sub-tree rooted at that non-empty child. Here's an example of this:
+
+```
+          R
+         / \
+        /   \
+       /     \
+      R0      R1
+     / \     / \
+    A   B   o   o 
+```
+
+where `o` stands for an empty leaf. This tree gets compacted to
+
+```
+          R'
+         / \
+        /   \
+       /     \
+      A       B
+```
```

Comment:
I wonder if we should replace this with a picture? This way, we could put the trees side by size: on the left no compaction, on the right with compaction, and an arrow going from left to right between them.

Reply:
![image](https://github.com/0xPolygonMiden/miden-base/assets/35031754/e1a81787-09d7-426c-95cf-5c21b82de1c8)


Maybe you can use that?

Reply:
Agree! I have added a drawing instead.